common/djangoapps/dark_lang/api.py

```diff
@@ -0,0 +1,199 @@
+    """
+    Support class and code for dark-launching languages.
+
+    This is configured by creating ``DarkLangConfig`` rows in the database,
+    using the django admin site.
+    """
+
+    @property
+    def released_langs(self):
+        """
+        Current list of released languages
+        """
+        language_options = DarkLangConfig.current().released_languages_list
+        if settings.LANGUAGE_CODE not in language_options:
+            language_options.append(settings.LANGUAGE_CODE)
+        return language_options
+
+    def process_darklang_request(self, request):
+        """
+        Prevent user from requesting un-released languages
+        """
+        if not DarkLangConfig.current().enabled:
+            return None
+
+        if RESET_SUBMIT in request.POST:
+            # Reset and clear the language preference
+            return 'Language reset to %s' % self.clear_preview_language(request)
+        if SET_LANGUAGE_SUBMIT in request.POST:
+            # Set the Preview Language
+            return self.set_preview_language(request)
+
+    def set_preview_language(self, request):
+        """
+        Attempt to set the Preview language
+        :param request: The Django Request Object
+        :return dark language setting code
+        """
+        if not DarkLangConfig.current().enabled:
+            return None
+
+        if LANGUAGE_INPUT_FIELD not in request.POST:
+            return None
+
+        preview_lang = request.POST[LANGUAGE_INPUT_FIELD]
+
+        if preview_lang == '':
+            return None
+
+        auth_user = request.user.is_authenticated()
+
+        # Set the session key to the requested preview lang
+        request.session[LANGUAGE_SESSION_KEY] = preview_lang
+
+        # Make sure that we set the requested preview lang as the dark lang preference for the
+        # user, so that the lang_pref middleware doesn't clobber away the dark lang preview.
+        if auth_user:
+            set_user_preference(request.user, DARK_LANGUAGE_KEY, preview_lang)
+        return 'Language set to %s' % preview_lang
+
+    def clear_preview_language(self, request):
+        """
+        Clears the dark language preview
+        :param request: The incoming Django Request
+        :return: The Response message
+        """
+        auth_user = request.user.is_authenticated()
+
+        # delete the session language key (if one is set)
+        if LANGUAGE_SESSION_KEY in request.session:
+            del request.session[LANGUAGE_SESSION_KEY]
+
+        user_pref = None
+        if auth_user:
+            # Reset user's dark lang preference to null
+            delete_user_preference(request.user, DARK_LANGUAGE_KEY)
+            # Get & set user's preferred language
+            user_pref = get_user_preference(request.user, LANGUAGE_KEY)
+            if user_pref:
+                request.session[LANGUAGE_SESSION_KEY] = user_pref
+        return user_pref
+
+    def fuzzy_match(self, lang_code):
```

Comment:
The naming convention of using a leading _ to indicate a private method is generally a good idea. I'd suggest using it for helper methods like this (and clear_preview_language, etc.) that are only pulled out for clarity.